compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -254,14 +263,17 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
     /// `to_backend_ty` must be the *non*-immediate backend type (so it will be
     /// `i8`, not `i1`, for `bool`-like types.)
     fn transmute_immediate(
+        &self,
         bx: &mut Bx,
         mut imm: Bx::Value,
         from_scalar: abi::Scalar,
+        from_backend_ty: Bx::Type,
         to_scalar: abi::Scalar,
         to_backend_ty: Bx::Type,
     ) -> Bx::Value {
         use abi::Primitive::*;
         imm = bx.from_immediate(imm);
+        self.assume_scalar_range(bx, imm, from_scalar, from_backend_ty);
```

Comment:
This is enough to let LLVM carry the information across the cast? Shouldn't we also do this again with the `from_backend_ty` after the bitcast?

Reply:
Hmm, I only added `-C no-prepopulate-passes` tests, didn't I.  I'll add more tests that actually run the optimizer to confirm that these do something.

As for the `bitcast`, the "bitcast" is actually nothing for things `char` -> `u32`, since they're the same `i32` type in LLVM.  So it's almost certainly fine for those cases in LLVM, though I'm unsure how cg_gcc's type system handles this.

Maybe I should try intersecting the ranges again.  I started doing that, but `WrappingRange ∧ WrappingRange` needs to be multiple `WrappingRange`s, so I'd given up on it...

@rustbot author

Reply:
Oh I'm totally fine with emitting multiple assumes that end up doing an intersection. I would just have expected you to assume the output range on the input and the input range on the output instead of just assuming each value's type's range